contrib/parseq-restli-client/src/main/java/com/linkedin/restli/client/ParSeqRestClient.java

```diff
@@ -194,13 +205,60 @@ static String generateTaskName(final Request<?> request) {
     }
   }
 
+  private <T> Task<Response<T>> withD2Timeout(final Task<Response<T>> task, ConfigValue<Long> timeout) {
+    if (timeout.getValue() != null && timeout.getValue() > 0) {
+      // transform TimeoutException to be compatible with current Task timeout exception message
+      String desc = timeout.getSource().map(src -> " src: " + src).orElse("");
+      String timeoutTaskName = "withTimeout " + Math.min(timeout.getValue(), Integer.MAX_VALUE)
+          + TimeUnitHelper.toString(TimeUnit.MILLISECONDS) + desc;
+      return task.transform(timeoutTaskName, (Try<Response<T>> tryGet) -> {
+        if (tryGet.isFailed()) {
+          if (tryGet.getError() instanceof TimeoutException) {
+            String timeoutExceptionMessage = "task: '" + task.getName() + "' " + timeoutTaskName;
+            return Failure.of(Exceptions.timeoutException(timeoutExceptionMessage));
+          } else {
+            return Failure.of(tryGet.getError());
+          }
+        } else {
+          return Success.of(tryGet.get());
+        }
+      });
+    } else {
+      return task;
+    }
+  }
+
   private <T> Task<Response<T>> createTask(final String name, final Request<T> request,
       final RequestContext requestContext, RequestConfig config) {
     LOGGER.debug("createTask, name: '{}', config: {}", name, config);
+    return createTaskWithTimeout(name, request, requestContext, config);
+  }
+
+  // reconcile timeout specified in requestContext with the given timeout
+  private void reconcileRequestTimeout(final RequestContext requestContext, Long timeout) {
+    Number contextTimeout = (Number) requestContext.getLocalAttr(R2Constants.REQUEST_TIMEOUT);
+    int timeoutBound = (contextTimeout != null) ? contextTimeout.intValue() : Integer.MAX_VALUE;
+    // d2 request timeout can only accept Integer timeout
+    requestContext.putLocalAttr(R2Constants.REQUEST_TIMEOUT, Math.min(timeout, timeoutBound));
+  }
+
+  private <T> Task<Response<T>> createTaskWithTimeout(final String name, final Request<T> request,
+      final RequestContext requestContext, RequestConfig config) {
+    ConfigValue<Long> timeout = config.getTimeoutMs();
+    if (timeout.getValue() != null && timeout.getValue() > 0 && _d2RequestTimeoutEnabled) {
```

Comment:
This if statement can be removed and reconcileRequestTimeout method invocation can be moved inside withD2Timeout inside "if (timeout.getValue() != null && timeout.getValue() > 0)" condition.

Reply:
hmm... reconcileRequest needs to update requestContext, which is used to create the parseq Task passed to withD2Timeout. Thus it will be too late to call reconcileRequestTimeout inside withD2Timeout?

Reply:
Yup, you are right.